src/cargo/util/toml/mod.rs

```diff
@@ -94,11 +105,71 @@ fn read_manifest_from_str(
 
     let mut unused = BTreeSet::new();
     let deserializer = toml::de::Deserializer::new(contents);
-    let manifest: manifest::TomlManifest = serde_ignored::deserialize(deserializer, |path| {
+    let manifest: manifest::TomlManifest = match serde_ignored::deserialize(deserializer, |path| {
         let mut key = String::new();
         stringify(&mut key, &path);
         unused.insert(key);
-    })?;
+    }) {
+        Ok(manifest) => manifest,
+        Err(e) => {
+            let Some(span) = e.span() else {
+                return Err(e.into());
+            };
+
+            let (line_num, column) = translate_position(contents, span.start);
+            let num_new_lines = contents[span.start..span.end]
+                .chars()
+                .filter(|c| c == &'\n')
+                .count()
+                - (contents[span.end - 1..span.end].ends_with('\n') as usize);
+            let lines = contents.lines();
+            let source: String = lines
+                .skip(line_num)
+                .take(num_new_lines + 1)
+                .enumerate()
+                .map(|(i, line)| {
+                    let mut line = line.to_string();
+                    // Add back the new line that was stripped by `.lines()
+                    if num_new_lines > i {
+                        line.push('\n');
+                    }
+                    line
+                })
+                .collect();
```

Comment:
Can't you just:
```rust
let source_start = contents[0..span.start].rfind('\n').map(|s| s + 1).unwrap_or(0);
let source_end = contents[span.end..].find('\n').unwrap_or(contents.len());
let source = &contents[source_start..source_end];
```

Reply:
Oops, I had a bug.  `source_end` needs `span.end` added to it.

Reply:
Thanks for this it works much better